modules/ml/src/main/java/org/apache/ignite/ml/nn/MLP.java

```diff
@@ -0,0 +1,503 @@
+        architecture = arch;
+        below = null;
+
+        initLayers(initializer);
+    }
+
+    /**
+     * Construct MLP from given architecture.
+     *
+     * @param arch Architecture.
+     */
+    public MLP(MLPArchitecture arch) {
+        layers = new ArrayList<>(arch.layersCount() + 1);
+        architecture = arch;
+        below = null;
+
+        initLayers(new RandomInitializer(new Random()));
+    }
+
+    /**
+     * Init layers parameters with initializer.
+     *
+     * @param initializer Parameters initializer.
+     */
+    private void initLayers(MLPInitializer initializer) {
+        int prevSize = architecture.inputSize();
+
+        for (int i = 1; i < architecture.layersCount(); i++) {
+            TransformationLayerArchitecture layerCfg = architecture.transformationLayerArchitecture(i);
+            int neuronsCnt = layerCfg.neuronsCount();
+            DenseLocalOnHeapMatrix weights = new DenseLocalOnHeapMatrix(neuronsCnt, prevSize);
+            initializer.initWeights(weights);
+            DenseLocalOnHeapVector biases = null;
+            if (layerCfg.hasBias()) {
+                biases = new DenseLocalOnHeapVector(neuronsCnt);
+                initializer.initBiases(biases);
+            }
+            layers.add(new MLPLayer(weights, biases));
+            prevSize = layerCfg.neuronsCount();
+        }
+    }
+
+    /**
+     * Create MLP from two MLPs: first stacked on second.
+     *
+     * @param above MLP to be above.
+     * @param below MLP to be below.
+     */
+    protected MLP(MLP above, MLP below) {
+        this.layers = above.layers;
+        this.architecture = above.architecture;
+        this.below = below;
+    }
+
+    /**
+     * Perform forward pass and return state of outputs of each layer.
+     *
+     * @param val Value to perform computation on.
+     * @return MLP state after computation.
+     */
+    public MLPState computeState(Matrix val) {
+        MLPState res = new MLPState(val);
+
+        forwardPass(val, res, true);
+
+        return res;
+    }
+
+    /**
+     * Perform forward pass and if needed write state of outputs of each layer.
+     *
+     * @param val Value to perform computation on.
+     * @param state State object to write state into.
+     * @param writeState Flag indicating need to write state.
+     */
+    public Matrix forwardPass(Matrix val, MLPState state, boolean writeState) {
+        Matrix res = val;
+
+        if (below != null)
+            res = below.forwardPass(val, state, writeState);
+
+        for (int i = 1; i < architecture.layersCount(); i++) {
+            MLPLayer curLayer = layers.get(i - 1);
+            res = curLayer.weights.times(res);
+
+            TransformationLayerArchitecture layerCfg = this.architecture.transformationLayerArchitecture(i);
+
+            if (layerCfg.hasBias()) {
+                ReplicatedVectorMatrix biasesMatrix = new ReplicatedVectorMatrix(biases(i), res.columnSize(), true);
```

Comment:
The name of type sounds crazy: VectorMatrix and also Replicated (sounds like something cached with Replicated regime)

It remainds me Dataset (specialized version)